src/AutoRest.CSharp.V3/Output/Builders/SerializationBuilder.cs

```diff
@@ -150,24 +150,70 @@ private static XmlObjectSerialization BuildXmlObjectSerialization(ObjectSchema o
                 schemaTypeReference, elements.ToArray(), attributes.ToArray(), embeddedArrays.ToArray());
         }
 
+        private static IEnumerable<JsonPropertySerialization> GetPropertySerializationsFromBag(PropertyBag propertyBag)
+        {
+            foreach (Property property in propertyBag.Properties)
+            {
+                yield return new JsonPropertySerialization(
+                    property.SerializedName,
+                    property.CSharpName(),
+                    BuildSerialization(property.Schema, property.IsNullable()),
+                    false);
+            }
+
+            foreach ((string name, PropertyBag innerBag) in propertyBag.Bag)
+            {
+                JsonPropertySerialization[] serializationProperties = GetPropertySerializationsFromBag(innerBag).ToArray();
+                JsonObjectSerialization objectSerialization = new JsonObjectSerialization(new SchemaTypeReference(null, false), serializationProperties, null);
+                yield return new JsonPropertySerialization(name, name, objectSerialization, true);
+            }
+        }
+
         private static JsonObjectSerialization BuildJsonObjectSerialization(ObjectSchema objectSchema, TypeReference schemaTypeReference, bool isNullable)
         {
             List<JsonPropertySerialization> serializationProperties = new List<JsonPropertySerialization>();
             foreach (var schema in EnumerateHierarchy(objectSchema))
             {
-                foreach (Property property in schema.Properties!)
+                PropertyBag propertyBag = new PropertyBag();
+                propertyBag.Properties.AddRange(schema.Properties ?? Array.Empty<Property>());
+                PopulatePropertyBag(propertyBag, 0);
+                serializationProperties.AddRange(GetPropertySerializationsFromBag(propertyBag));
+            }
+
+            return new JsonObjectSerialization(schemaTypeReference, serializationProperties.ToArray(), CreateAdditionalProperties(objectSchema));
+        }
+
+        private class PropertyBag
+        {
+            public Dictionary<string, PropertyBag> Bag { get; } = new Dictionary<string, PropertyBag>();
+            public List<Property> Properties { get; } = new List<Property>();
+        }
+
+        private static void PopulatePropertyBag(PropertyBag propertyBag, int depthIndex)
+        {
+            foreach (Property property in propertyBag.Properties.ToArray())
+            {
+                int flattenNamesCount = property.FlattenedNames?.Count ?? 0;
+                if (flattenNamesCount == 0 || depthIndex >= flattenNamesCount - 1)
```

Comment:
Can't you merge these two?

Reply:
Fixed. Removed the first condition as it was redundant. If there are no flatten names, the `count - 1` will be -1 and the first depthIndex is 0. Therefore, 0 is greater or equal to -1, which means will continue properly.